drivers/wifi/esp32/src/esp_wifi_drv.c

```diff
@@ -732,24 +931,53 @@ static const struct wifi_mgmt_ops esp32_wifi_mgmt = {
 	.scan		   = esp32_wifi_scan,
 	.connect	   = esp32_wifi_connect,
 	.disconnect	   = esp32_wifi_disconnect,
+#ifndef CONFIG_ESP32_WIFI_APSTA_MODE
 	.ap_enable	   = esp32_wifi_ap_enable,
 	.ap_disable	   = esp32_wifi_ap_disable,
+#endif
 	.iface_status	   = esp32_wifi_status,
 #if defined(CONFIG_NET_STATISTICS_WIFI)
 	.get_stats	   = esp32_wifi_stats,
 #endif
 };
 
+#ifdef CONFIG_ESP32_WIFI_APSTA_MODE
+static const struct wifi_mgmt_ops esp32_wifi_ap_mode_mgmt = {
+	.ap_enable = esp32_wifi_ap_enable,
+	.ap_disable = esp32_wifi_ap_disable,
+	.iface_status = esp32_wifi_status,
+#if defined(CONFIG_NET_STATISTICS_WIFI)
+	.get_stats = esp32_wifi_stats,
+#endif
+};
+#endif
```

Comment:
Why do we need a separate ops? Based on the interface, can't the existing ops handle standalone AP and concurrent AP and STA modes.

Reply:
because to not allow AP related functions with STA interfaces and STA related functions with AP interfaces. If AP is enabled on a STA interface, conflicts could arise since each interface can only have one IP address associated with it.

Reply:
So, IIUC,

With `CONFIG_ESP32_WIFI_APSTA_MODE=n`: single interface that supports STA or AP
With `CONFIG_ESP32_WIFI_APSTA_MODE=y`: first interface STA only and second interface AP only

This works, but if you want a generic implementation, then you can allow both interfaces to be either STA or AP, let user choose that and we don't even need `CONFIG_ESP32_WIFI_APSTA_MODE`, of course it needs some intelligence/state, something like this (crude pseudo code), WDYT?

```
iface_is_idle:
   return !AP_UP || !STA_CONNECTED

esp32_wifi_ap_enable:
    if !iface_is_idle:
       return -1
   if scanning:
      abort

connect:
    if !iface_is_idle:
       return -1   
```

Reply:
> CONFIG_ESP32_WIFI_APSTA_MODE still be needed because if 2nd wifi node is not defined in .overlay file, compiler will through error on defining 2nd net_if. CONFIG_ESP32_WIFI_APSTA_MODE will allow to compile the code successfully if only one wi-fi node is defined in .overlay. Following is the block of code I'm concern about.

Sure, while trying to register the interfaces you can use that Kconfig to control, I was only referring to the usage in the rest of the code.

> Secondly, How can I check which net_if has called which function? for-example, in following callback function, how can we get net_if from device? esp32_wifi_iface is defined globally.

`struct net_if *iface = net_if_lookup_by_dev(dev);`

This way, you don't need to rely on any global ifaces, just check the current state of the interfaces and allow or accept the operation.

Reply:
The following functions are dependent on global interfaces. Can you suggest how can we determine the ifaces in these function? Because i didn't find a way of of determining which interfaces needs to be handle in following functions. Maybe you can give some idea.
```
static esp_err_t eth_esp32_rx(void *buffer, uint16_t len, void *eb);
static void esp_wifi_handle_sta_connect_event(void *event_data);
static void esp_wifi_handle_sta_disconnect_event(void *event_data);
```
The following event handler also depends on global iface.
```
void esp_wifi_event_handler(const char *event_base, int32_t event_id, void *event_data,
			    size_t event_data_size, uint32_t ticks_to_wait)
{
	LOG_DBG("Wi-Fi event: %d", event_id);
	switch (event_id) {
	case WIFI_EVENT_STA_START:
		esp32_data.state = ESP32_STA_STARTED;
		net_eth_carrier_on(esp32_wifi_iface);
		break;
	case WIFI_EVENT_STA_STOP:
		esp32_data.state = ESP32_STA_STOPPED;
		net_eth_carrier_off(esp32_wifi_iface);
		break;
	case WIFI_EVENT_STA_CONNECTED:
		esp_wifi_handle_sta_connect_event(event_data);
		break;
	case WIFI_EVENT_STA_DISCONNECTED:
		esp_wifi_handle_sta_disconnect_event(event_data);
		break;
	case WIFI_EVENT_SCAN_DONE:
		scan_done_handler();
		break;
}
```

Reply:
It really depends on driver author how this is managed, but I can share how we do it in `nrf70` driver.  See https://github.com/krish2718/zephyr/blob/add_7002ek/drivers/wifi/nrfwifi/src/net_if.c#L191 for details.

* All internal functions are based on `vif_index` (this might not  be the case in ESP)
* `NET_DEVICE_DT_INST_DEFINE` passes the `vif_ctx` as data, this wasy `dev/net_if` can be mapped to `vif_ctx` (driver struct) and vice-versa.
* Driver-Stack: `vif_index` -> `vif_ctx` -> `net_if`
* Stack->Driver: `dev` -> `(vif_ctx)dev->data`

Instead of global this cleans implement per-iface data structure that can be used to store all per-iface/VIF information.

Reply:
> > CONFIG_ESP32_WIFI_APSTA_MODE still be needed because if 2nd wifi node is not defined in .overlay file, compiler will through error on defining 2nd net_if. CONFIG_ESP32_WIFI_APSTA_MODE will allow to compile the code successfully if only one wi-fi node is defined in .overlay. Following is the block of code I'm concern about.
> 
> Sure, while trying to register the interfaces you can use that Kconfig to control, I was only referring to the usage in the rest of the code.
> 
> > Secondly, How can I check which net_if has called which function? for-example, in following callback function, how can we get net_if from device? esp32_wifi_iface is defined globally.
> 
> `struct net_if *iface = net_if_lookup_by_dev(dev);`
> 
> This way, you don't need to rely on any global ifaces, just check the current state of the interfaces and allow or accept the operation.

Hi Krish2718,
Thank you for your valueable inputs. I am also working with @zin-HXQ on this driver. Actualy we were having issues enabling both modes of wifi simultaneously with single netif interface. To fix this issue, we need two netif as it was also done by espressif in `esp-idf platform`. That's the reason to keep two ops in the code. 

On keeping iface global, we followed the same code technique as was used by `esp_wifi_drv` already. Further, there is no `dev` or `iface` parameters in most of the functions due to which we are unable to check which interface is sending these callback. The callback `esp_wifi_event_handler` is required by espressif `hal` api and we can not change it.

Reply:
> Actualy we were having issues enabling both modes of wifi simultaneously with single netif interface. To fix this issue, we need two netif as it was also done by espressif in esp-idf platform. That's the reason to keep two ops in the code.

Sure, we do need two separate interfaces for AP + STA concurrency. But my comment is using the same set of ops for both interfaces and then using dev/iface handle them accordingly.

> On keeping iface global, we followed the same code technique as was used by esp_wifi_drv already. Further, there is no dev or iface parameters in most of the functions due to which we are unable to check which interface is sending these callback. The callback esp_wifi_event_handler is required by espressif hal api and we can not change it.

Understood, so, globals are necessary, but can this be done using single set of ops?

Reply:
Yes, we have done it with single ops. We are testing it now.

Reply:
We have changed the dependency on single ops. and tried to avoid global iface where we can.

Reply:
Thanks, this is looking much cleaner.